### PR TITLE
New documentation for Java

### PR DESCRIPTION
**Pull request by valtandor**
François, Wolfgang: Could you please look at this new document and verify its correctness, especially the example submission scripts for Fitzroy and Beatrice?

### CODE REVIEW

File: Java.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+Java is a computer programming language that is concurrent, class-based,
+object-oriented, and specifically designed to have as few implementation
+dependencies as possible. It is intended to let application developers "write
+once, run anywhere" (WORA), meaning that code that runs on one platform does
+not need to be recompiled to run on another. Java applications are typically
+compiled to bytecode (class file) that can run on any Java virtual machine
+(JVM) regardless of computer architecture. Java is, as of 2012, one of the most
+popular programming languages in use, particularly for client-server web
+applications, with a reported 9 million developers.[10][11] Java was originally
```
Comment (kiwifb) [inline Java.md]: [10] and [11] were probably references in the original bits you copied. I guess the easiest thing to do is remove them unless you want to replace them by links.

File: Java.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	#@ class = p7linux
+	#@ wall_clock_limit = 01:00:00
+	#@ job_type = parallel
+	#@ node = 1
+	#@ tasks_per_node = 1
+	#@ parallel_threads = 8
+	#@ task_affinity = core(8)
+	#@ rset = rset_mcm_affinity
+	#@ cpus_per_core = 1
+	#@ resources = ConsumableMemory(32gb)
```
Comment (kiwifb) [inline Java.md]: We don't set memory resources on power7. I believe we only ever used `resources` for ansys licence on aix. Not sure if it is used on fitzroy either. I believe you need to set particular settings in loadleveler and we don't here and I haven't heard anything about fitzroy.

Comment (valtandor) [inline Java.md]: How do you control how much memory a task uses then? And how would the user know what value to set -Xmx to?

File: Java.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	#@ initialdir = /hpcf/working/nesi99999/MyJavaJob
```
Comment (kiwifb) [inline Java.md]: We don't usually set `initialdir` but let's put something reasonable on power7 like `/hpc/scratch/nesi99999/MyJavaJob`.

File: Java.md (hunk 2 above)
Comment (kiwifb) [inline Java.md]: We don't impose control on the memory. The nodes have 128GB of memory and you should set the memory with the idea of 4GB per core.

Comment (valtandor) [inline Java.md]: OK, the resources line has been removed.

**Comment by kiwifb** [comment]
I think Wolfgang should have a closer look at the fitzroy bit. I don't know if they do memory management there either.

**Comment by valtandor** [comment]
@tinyendian would you be so kind as to take a look at this pull request?

**Comment by tinyendian** [comment]
Hi, I'm not aware of memory resources being usually set on Fitzroy (I normally don't set these), but I'll need to check this with our HPC admins. It would certainly make a lot of sense to get people to set limitations for Java if it is notorious for causing problems with memory management, to avoid problems with incorrect setups, as it happened last week.

**Comment by tinyendian** [comment]
Fitzroy memory setup:
* setting "resources = ConsumableMemory(32gb)" only acts as a selection criterion for LoadLeveler to choose the right node according to available memory (in particular if node usage is shared), but it does not restrict runtime usage of memory in any way
* a Java job can still use more memory than what is set here, and the offender (or some other, unrelated tasks that are running on the node) will only be killed by AIX if general memory limits are exceeded
* we should thus encourage users to set the Xmx flag, to avoid problems if leaky Java codes are run
* standard Fitzroy nodes have up to 55 GB of memory available if used with setting "node_usage = not_shared" (which is the default), which we could add as a guideline comment in the script

In my opinion, using "ConsumableMemory" is thus not necessary or useful, unless a user wants share nodes specifically, which most users probably won't want to do - in particular since Fitzroy is meant to be used mainly for bigger parallel jobs.

The setup script looks good to me otherwise.

**Comment by valtandor** [comment]
OK. For what it's worth, the documentation at https://teamwork.niwa.co.nz/display/HPCF/Fitzroy+User+guide#FitzroyUserguide-5.1IntroductiontoLoadLeveler doesn't say anything about "node_usage", nor that in the majority of cases the resources directive doesn't accomplish anything much.

**Comment by tinyendian** [comment]
It's true that the documentation could be improved in many places. Although "node_usage" is something that should be of little concern for most users - it is only really relevant if the queue fills up, and if there is a number of sufficiently small jobs that don't have too high demands for node resources and could thus run together on the same node.

**Comment by valtandor** [comment]
OK, I've removed the resources directive for the example Fitzroy submission script.

**Pull request merged**